drivers/ethernet/eth_nxp_s32_gmac.c

```diff
@@ -0,0 +1,777 @@
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/net_pkt.h>
+#include <ethernet/eth_stats.h>
+#include <soc.h>
+
+#include <Gmac_Ip.h>
+#include <Gmac_Ip_Hw_Access.h>
+#include <Gmac_Ip_Irq.h>
+#include <Clock_Ip.h>
+
+#include "eth.h"
+
+#define ETH_NXP_S32_BUF_TIMEOUT		K_MSEC(20)
+#define ETH_NXP_S32_DMA_TX_TIMEOUT	K_MSEC(20)
+
+#define ETH_NXP_S32_MAC_ADDR_LEN 6U
+
+#define FREESCALE_OUI_B0 0x00
+#define FREESCALE_OUI_B1 0x04
+#define FREESCALE_OUI_B2 0x9f
+
+struct eth_nxp_s32_config {
+	uint8_t instance;
+	uint8_t tx_ring_idx;
+	uint8_t rx_ring_idx;
+	uint32_t rx_irq;
+	uint32_t tx_irq;
+	void (*do_config)(void);
+	const struct pinctrl_dev_config *pincfg;
+
+	const Gmac_CtrlConfigType ctrl_cfg;
+	GMAC_Type *base;
+};
+
+struct eth_nxp_s32_data {
+	struct net_if *iface;
+	uint8_t mac_addr[ETH_NXP_S32_MAC_ADDR_LEN];
+	struct k_mutex tx_mutex;
+	struct k_sem rx_sem;
+	struct k_sem tx_sem;
+	struct k_thread rx_thread;
+
+	K_KERNEL_STACK_MEMBER(rx_thread_stack, CONFIG_ETH_NXP_S32_RX_THREAD_STACK_SIZE);
+};
+
+static void eth_nxp_s32_rx_thread(void *arg1, void *unused1, void *unused2);
+
+static inline struct net_if *get_iface(struct eth_nxp_s32_data *ctx, uint16_t vlan_tag)
+{
+#if defined(CONFIG_NET_VLAN)
+	struct net_if *iface;
+
+	iface = net_eth_get_vlan_iface(ctx->iface, vlan_tag);
+	if (!iface) {
+		return ctx->iface;
+	}
+
+	return iface;
+#else
+	ARG_UNUSED(vlan_tag);
+
+	return ctx->iface;
+#endif
+}
+
+#if defined(CONFIG_SOC_PART_NUMBER_S32K3)
+static int select_phy_interface(Gmac_Ip_MiiModeType mode)
+{
+	uint32_t regval;
+
+	switch (mode) {
+	case GMAC_MII_MODE:
+		regval = DCM_GPR_DCMRWF1_EMAC_CONF_SEL(0U);
+		break;
+	case GMAC_RMII_MODE:
+		regval = DCM_GPR_DCMRWF1_EMAC_CONF_SEL(2U);
+		break;
+#if (FEATURE_GMAC_RGMII_EN == 1U)
+	case GMAC_RGMII_MODE:
+		regval = DCM_GPR_DCMRWF1_EMAC_CONF_SEL(1U);
+		break;
+#endif
+	default:
+		return -EINVAL;
+	}
+
+	IP_DCM_GPR->DCMRWF1 = (IP_DCM_GPR->DCMRWF1 & ~DCM_GPR_DCMRWF1_EMAC_CONF_SEL_MASK) | regval;
+
+	return 0;
+}
+#else
+#error "SoC not supported"
+#endif /* CONFIG_SOC_PART_NUMBER_S32K3 */
+
+static int eth_nxp_s32_init(const struct device *dev)
+{
+	const struct eth_nxp_s32_config *cfg = dev->config;
+	struct eth_nxp_s32_data *ctx = dev->data;
+	Gmac_Ip_StatusType mac_status;
+	Clock_Ip_StatusType clk_status;
+	int err;
+
+	err = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err != 0) {
+		return err;
+	}
+
+	/*
+	 * Currently, clock control shim driver does not support configuring clock
+	 * muxes individually, so use the HAL directly.
+	 */
+	clk_status = Clock_Ip_Init(&Clock_Ip_aClockConfig[CONFIG_ETH_NXP_S32_CLOCK_CONFIG_IDX]);
+	if (clk_status != CLOCK_IP_SUCCESS) {
+		LOG_ERR("Failed to configure clocks (%d)", clk_status);
+		return -EIO;
+	}
+
+	/*
+	 * PHY mode selection must be done before the controller is reset,
+	 * because the interface type is latched at controller's reset
+	 */
+	err = select_phy_interface(cfg->ctrl_cfg.Gmac_pCtrlConfig->MiiMode);
```

Comment:
Should this be wrapped with the check for `CONFIG_SOC_PART_NUMBER_S32K3`

Reply:
The phy interface selection must be done always independently of the SoC. Atm the only implementation provided is for K344, but later can be extended when more devices are supported.